File: source/extensions/omni.cae.index/plugins/omni.cae.index/Extension.cpp
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
//
// NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
// property and proprietary rights in and to this material, related
// documentation and any modifications thereto. Any use, reproduction,
// disclosure or distribution of this material and related documentation
// without an express license agreement from NVIDIA CORPORATION or
// its affiliates is strictly prohibited.

#define CARB_EXPORTS

#include "Importers.h"

#include <carb/PluginUtils.h>
#include <carb/logging/Log.h>

#include <nv/index/app/application_layer/iapplication_layer.h>
#include <nv/index/app/application_layer/iimporter_manager.h>
#include <nv/index/app/application_layer/iinterface_factory_manager.h>
#include <nv/index/iindex.h>
#include <omni/ext/IExt.h>
#include <rtx/index/IndexInstance.h>

#include <cassert>

#define EXTENSION_NAME "omni.cae.index.plugin"

using namespace carb;
CARB_PLUGIN_IMPL_DEPS(rtx::index::IndexInstance, carb::logging::ILogging);

class Extension : public omni::ext::IExt
{
    std::string m_extensionId;
    nv::index::IIndex* m_iindex = nullptr;
    nv::index::app::IApplication_layer* m_application_layer = nullptr;
    mi::base::Handle<nv::index::app::IImporter_manager> m_importer_manager;
    mi::base::Handle<nv::index::app::IInterface_factory_manager> m_interface_manager;
    mi::base::Handle<omni::cae::index::ImporterFactory> m_importer_factory;
    mi::base::Handle<omni::cae::index::InterfaceFactory> m_interface_factory;

public:
    void onStartup(const char* extId) override
    {
        this->m_extensionId = extId;
        auto* ifaceIndex = carb::getFramework()->acquireInterface<rtx::index::IndexInstance>();
        m_iindex = ifaceIndex->getInterface<nv::index::IIndex>();
        m_application_layer = ifaceIndex->getInterface<nv::index::app::IApplication_layer>();

        // Retrieve and store the factory for creating Python-based importers
        m_importer_manager = m_application_layer->get_api_component<nv::index::app::IImporter_manager>();
        if (m_importer_manager)
        {
            m_importer_factory = new omni::cae::index::ImporterFactory(m_iindex, m_application_layer);
            if (m_importer_manager->register_importer_factory(m_importer_factory.get()) == 0)
            {
                CARB_LOG_INFO("Index importers for namespace '%s' registered successfully",
                              m_importer_factory->get_namespace_string());
            }
            else
            {
                CARB_LOG_ERROR("Failed to register omni::cae::index::ImporterFactory.");
            }
        }
        else
        {
            CARB_LOG_ERROR("IImporter_manager is not available.");
        }

        m_interface_manager = m_application_layer->get_api_component<nv::index::app::IInterface_factory_manager>();
        assert(m_interface_manager.is_valid_interface());
        m_interface_factory = new omni::cae::index::InterfaceFactory();
        if (m_interface_manager->register_iinterface_factory(m_interface_factory.get()) == 0)
        {
            CARB_LOG_INFO("Index interfaces for namespace '%s' registered successfully",
                          m_interface_factory->get_namespace_string());
        }
        else
        {
            CARB_LOG_ERROR("Failed to register omni::cae::index::InterfaceFactory.");
        }
    }

    void onShutdown() override
    {
        if (m_importer_manager && m_importer_factory)
        {
            m_importer_manager->unregister_importer_factory(m_importer_factory->get_namespace_string());
        }
    }
};

const struct carb::PluginImplDesc kPluginImpl = { EXTENSION_NAME, "Omni CAE IndeX Plugin.", "NVIDIA",
                                                  carb::PluginHotReload::eDisabled, "dev" };
CARB_PLUGIN_IMPL(kPluginImpl, Extension);
void fillInterface(Extension& iface)
{
}

File: source/extensions/omni.cae.property.bundle/python/extension.py
# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
#
# NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
# property and proprietary rights in and to this material, related
# documentation and any modifications thereto. Any use, reproduction,
# disclosure or distribution of this material and related documentation
# without an express license agreement from NVIDIA CORPORATION or
# its affiliates is strictly prohibited.

__all__ = ["Extension"]

from omni.ext import IExt
from omni.kit.property.bundle import GeomPrimSchemeDelegate
from omni.kit.window.property import get_window

from .property_widget import (
    CaeGeomPrimSchemeDelegate,
    CaePropertiesWidget,
    OmniSciPropertiesWidget,
    RtwtPropertiesWidget,
)


class Extension(IExt):

    def on_startup(self, ext_id):
        if property_window := get_window():
            property_window.register_widget("prim", "cae", CaePropertiesWidget("CAE"))
            property_window.register_widget("prim", "rtwt", RtwtPropertiesWidget("RTWT"))
            property_window.register_widget("prim", "omni_sci", OmniSciPropertiesWidget("Omni Scientific"))
            property_window.register_scheme_delegate("prim", "xformable_prim", CaeGeomPrimSchemeDelegate())

    def on_shutdown(self):
        if property_window := get_window():
            property_window.unregister_widget("prim", "omni_sci")
            property_window.unregister_widget("prim", "rtwt")
            property_window.unregister_widget("prim", "cae")
            # restore the default GeomPrimSchemeDelegate
            property_window.register_scheme_delegate("prim", "xformable_prim", GeomPrimSchemeDelegate())
